compiler/rustc_infer/src/infer/error_reporting/mod.rs

```diff
@@ -2187,11 +2190,63 @@ impl<'tcx> TypeErrCtxt<'_, 'tcx> {
                 .unwrap_or("'lt".to_string())
         };
 
-        let add_lt_sugg = generics
-            .params
-            .first()
-            .and_then(|param| param.def_id.as_local())
-            .map(|def_id| (self.tcx.def_span(def_id).shrink_to_lo(), format!("{}, ", new_lt)));
+        let mut add_lt_suggs: Vec<Option<_>> = vec![];
+        if is_synthetic {
+            if let Some(ast_generics) = ast_generics {
+                if ast_generics.params.iter().any(|p| {
+                    matches!(
+                        p.kind,
+                        hir::GenericParamKind::Lifetime { kind: hir::LifetimeParamKind::Explicit }
+                    )
+                })
+                && let Some(def_id) = ast_generics.params.first().and_then(|param| Some(param.def_id)) {
+                    add_lt_suggs.push(Some((
+                        self.tcx.def_span(def_id).shrink_to_lo(),
+                        format!("{new_lt}, "),
+                    )));
+                } else {
+                    add_lt_suggs
+                        .push(Some((ast_generics.span.shrink_to_hi(), format!("<{new_lt}>"))));
+                }
+            }
+        } else {
+            if let Some(def_id) = generics.params.first().and_then(|param| param.def_id.as_local())
```

Comment:
This could be two nested `if let` or something like 

```suggestion
            if let [param, ..] = generics.params && let Some(def_id) = param.def_id.as_local()
```

Reply:
Changed.